src/main/java/pro/sky/stream_api_optional_2_8/service/EmployeeServiceMapImpl.java

```diff
@@ -0,0 +1,116 @@
+
+import static org.apache.commons.lang3.StringUtils.isAlpha;
+import static pro.sky.stream_api_optional_2_8.utils.Constants.*;
+
+
+@Service
+public class EmployeeServiceMapImpl implements EmployeeService {
+
+    private final Map<String, Employee> employees = new HashMap<>(EMPLOYEE_AMOUNT);
+
+    @Override
+    public Employee getEmployeeWithMinSalary(Integer departmentId) {
+        return getAllEmployees().stream()
+                .filter((employee) -> employee.getDepartment().equals(departmentId))
+                .min(Comparator.comparing(Employee::getSalary))
+                .orElseThrow(() -> new EmployeeNotFoundException("Employee not found", HttpStatus.BAD_REQUEST));
+    }
+
+    @Override
+    public List<Employee> getAllEmployees(Integer departmentId) {
+        return getAllEmployees().stream()
+                .filter((employee) -> employee.getDepartment().equals(departmentId))
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public Map<Integer, List<Employee>> getAllEmployeesByDepartments() {
+        return getAllEmployees().stream()
+                .collect(Collectors.groupingBy((Employee::getDepartment)));
+    }
+
+
+    @Override
+    public Employee getEmployeeWithMaxSalary(Integer departmentId) {
+        return getAllEmployees().stream()
+                .filter((employee) -> employee.getDepartment().equals(departmentId))
+                .max(Comparator.comparing(Employee::getSalary))
+                .orElseThrow(() -> new EmployeeNotFoundException("Employee not found", HttpStatus.BAD_REQUEST));
+    }
+
+    @Override
+    public Employee addEmployee(String name, Integer department, Integer salary) {
+        validateNames(name);
+
+        Employee newEmployee = new Employee(name, department, salary);
+        getAllEmployees().stream()
+                .filter(emp -> emp.equals(newEmployee))
+                .findAny()
+                .ifPresent(emp -> {
+                    throw new EmployeeAlreadyAddedException("Employee " + name +
+                            " in department " + department +
+                            " with salary " + salary +
+                            " is already added",HttpStatus.BAD_REQUEST);
+                });
+        if (getAllEmployees().size() >= EMPLOYEE_AMOUNT) {
+            throw new ArrayIsFullException("Array is full",HttpStatus.BAD_REQUEST);
+        }
+
+        employees.put(name, newEmployee);
+        return newEmployee;
+    }
+
+    private void validateNames(String... names) {
+        for (String name : names) {
+            if (!isAlpha(name)) {
+                throw new InvalidNameException("Invalid name", HttpStatus.BAD_REQUEST);
+            }
+        }
+    }
+
+    @Override
+    public Employee removeEmployee(String name, Integer department, Integer salary) {
+        Employee employeeToRemove = new Employee(name, department, salary);
+        if (!employees.get(name).equals(employeeToRemove)) {
```

Comment:
для удаления тоже нужно только имя, отдел и ЗП не нужны. Надо в мапе проверить по имени просто и всё, никакой временный костыльный объект не нужен
```suggestion
        if (!employees.containsKey(name)) {
```